### PR TITLE
Allow reserved keywords in import/export when safe to do so

### LINKED ISSUE #828: default keyword not allowed as import name from sourced module declarations
Flow throws the syntax error "Unexpected token default" on the following input source code:

```javascript
export { default as foo } from "./bar";
```

```
$ cat test.js
export { default as foo } from "./bar";
$ flow ast test.js
{"errors":[{"loc":{"source":null,"start":{"line":1,"column":9},"end":{"line":1,"column":16}},"message":"Unexpected token default"}],"tokens":[],"type":"Program","loc":{"source":null,"start":{"line":1,"column":0},"end":{"line":1,"column":39}},"range":[0,39],"body":[{"type":"ExportDeclaration","loc":{"source":null,"start":{"line":1,"column":0},"end":{"line":1,"column":39}},"range":[0,39],"default":false,"declaration":null,"specifiers":[{"type":"ExportSpecifier","loc":{"source":null,"start":{"line":1,"column":9},"end":{"line":1,"column":23}},"range":[9,23],"id":{"type":"Identifier","loc":{"source":null,"start":{"line":1,"column":9},"end":{"line":1,"column":16}},"range":[9,16],"name":"default","typeAnnotation":null,"optional":false},"name":{"type":"Identifier","loc":{"source":null,"start":{"line":1,"column":20},"end":{"line":1,"column":23}},"range":[20,23],"name":"foo","typeAnnotation":null,"optional":false}}],"source":{"type":"Literal","loc":{"source":null,"start":{"line":1,"column":31},"end":{"line":1,"column":38}},"range":[31,38],"value":"./bar","raw":"\"./bar\""},"exportKind":"value"}],"comments":[]}
```

### PR DESCRIPTION
**Pull request by samwgoldman**
If we are doing a delegate export, the identifiers in the export
specifiers come from and go to object keys, so they follow the same
rules as object keys. If there is no "from" in the export statement, the
identifiers come from bindings, and thus follow those rules.

Import statements can rebind identifiers that come from object keys to
valid binding identifiers. If there is an "as" in the import binding
list item, the left-hand identifier can be a reserved keyword.

In order to allow the parse with the option of raising later, I modified
the signature of `identifier_or_reserved_keyword` to return the parsed
identifier along with the error that would have been raised. I factored
out the logic to determine this error from `error_unexpected`.

Fixes #828, #823

### CODE REVIEW

File: src/parser/parser_flow.ml [hunk 1]
```diff
@@ -1846,9 +1848,11 @@ end = struct
...
      *)
     and identifier_or_reserved_keyword env =
-      match Peek.token env with
-      | T_IDENTIFIER -> Parse.identifier env
-      | _ -> (match Peek.token env with
+      let { lex_token; lex_value; lex_loc; _ } = lookahead env in
+      match lex_token with
+      | T_IDENTIFIER -> Parse.identifier env, None
+      | _ ->
+        let err = match Peek.token env with
```
Comment (gabelevi) [inline src/parser/parser_flow.ml]: You could `s/Peek.token env/lex_token/` right?

File: src/parser/parser_flow.ml [hunk 2]
```diff
@@ -105,6 +105,16 @@ let rec filter_duplicate_errors acc = function
     filter_duplicate_errors acc rl
 | x :: rl -> filter_duplicate_errors (x :: acc) rl
 
+let unexpected_error = function
```
Comment (gabelevi) [inline src/parser/parser_flow.ml]: The naming is a little unfortunate with `error_unexpected` right below. Maybe `make_unexpected_error` or `get_unexpected_error` or something to help the dyslexics among us (me)?

Comment (samwgoldman) [inline src/parser/parser_flow.ml]: Haha. I was definitely being a bit cute here, but I can see how this is confusing. `get_unexpected_error` works for me.

File: src/parser/test/hardcoded_tests.js [hunk 3]
```diff
@@ -2590,5 +2590,22 @@ module.exports = {
         }
       }
     }
+  },
+  'Valid export/import reserved words': {
```
Comment (gabelevi) [inline src/parser/test/hardcoded_tests.js]: Wow this is complicated! Took me a few minutes to grok what was and wasn't allowed and why!

**Pull request closed without merging**